src/shared/addons/addonreplacer.cpp

```diff
@@ -0,0 +1,149 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "addonreplacer.h"
+
+#include <QDir>
+#include <QFile>
+#include <QFileInfo>
+#include <QJsonArray>
+#include <QJsonObject>
+#include <QQmlEngine>
+#include <QScopeGuard>
+
+#include "leakdetector.h"
+#include "logger.h"
+#include "qmlengineholder.h"
+#include "resourceloader.h"
+
+namespace {
+Logger logger("AddonReplacer");
+
+QUrl normalizeUrlScheme(const QUrl& url, const QString& response) {
+  // QT is a strange framework. Sometimes the files are loaded using 'qrc:'
+  // scheme, sometimes, just ':'. Let's normalize the final URL based on the
+  // incoming one.
+
+  if (url.scheme() == "qrc" && response.startsWith(':')) {
+    return QString("qrc%1").arg(response);
+  }
+  return response;
+}
+}  // namespace
+
+// static
+Addon* AddonReplacer::create(QObject* parent, const QString& manifestFileName,
+                             const QString& id, const QString& name,
+                             const QJsonObject& obj) {
+  QJsonObject replacerObj = obj["replacer"].toObject();
+
+  QJsonArray urls = replacerObj["urls"].toArray();
+  if (urls.isEmpty()) {
+    logger.warning() << "Empty URLs";
+    return nullptr;
+  }
+
+  AddonReplacer* replacer =
+      new AddonReplacer(parent, manifestFileName, id, name);
+  auto guard = qScopeGuard([&] { replacer->deleteLater(); });
+
+  QFileInfo manifestFileInfo(manifestFileName);
+  QDir addonPath = manifestFileInfo.dir();
+
+  for (const QJsonValue& urlValue : urls) {
+    QJsonObject urlObj = urlValue.toObject();
+
+    QString urlRequest = urlObj["request"].toString();
+    QString urlResponse = urlObj["response"].toString();
+    QString urlResponseFile(addonPath.filePath(urlResponse));
+
+    if (urlRequest.isEmpty() || urlResponse.isEmpty()) {
+      logger.warning() << "Invalid URL content";
+      return nullptr;
+    }
+
+    QString type = urlObj["type"].toString();
+    if (type.isEmpty() || type == "file") {
+      if (!QFile::exists(urlResponseFile)) {
```

Comment:
Should this also check that this in fact a `qrc://`? 
Otherwise what could stop me from doing `urlResponseFile="../../../../user_controlled_evil.qml"`